### PR TITLE
k-planes implementation

### PR DESCRIPTION
**Pull request by hturki**
This should be a pretty faithful implementation of the official K-Planes implementation at https://github.com/sarafridov/K-Planes. 

It can load models trained with the official implementation and gives comparable PSNR metrics when trained from scratch (31.69 vs 31.86 db in PSNR)

### CODE REVIEW

**Comment by hturki** [comment]
Update - the original difference in PSNR was due to a dataloader bug in the original implementation (https://github.com/sarafridov/K-Planes/issues/11). Both implementations now give comparable PSNR values and I've updated the PR description accordingly.

File: nerfstudio/data/dataparsers/blender_dataparser.py [hunk 1]
```diff
@@ -88,7 +88,10 @@ def _generate_dataparser_outputs(self, split="train"):
 
         # in x,y,z order
         camera_to_world[..., 3] *= self.scale_factor
-        scene_box = SceneBox(aabb=torch.tensor([[-1.5, -1.5, -1.5], [1.5, 1.5, 1.5]], dtype=torch.float32))
+        radius = 1.3 if "ship" not in str(self.data) else 1.5
```
Comment (tancik) [inline nerfstudio/data/dataparsers/blender_dataparser.py:91]: Remove special logic per scene.

**Review by tancik — COMMENTED** [review]
Thanks for reimplementing k-planes.

I left some initial comments. One broader one is that I would like to avoid adding the alpha compositing in this PR. It's mostly tangential to k-planes and touches many base files.

File: nerfstudio/fields/kplanes_field.py [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+try:
+    import tinycudann as tcnn
+except ImportError:
+    # tinycudann module doesn't exist
+    pass
+
+CONSOLE = Console(width=120)
+
+
+def init_grid_param(grid_nd: int, in_dim: int, out_dim: int, reso: List[int], a: float = 0.1, b: float = 0.5):
```
Comment (Giodiro) [inline nerfstudio/fields/kplanes_field.py:46]: Here `grid_nd` is always 2, it would be cleaner to remove the configuration option and other references.

Comment (Giodiro) [reply nerfstudio/fields/kplanes_field.py:46]: Also `in_dim` can be easily inferred from `reso` (`in_dim = len(reso)`), best to remove it from the arguments.

File: nerfstudio/fields/kplanes_field.py [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+    interp = interp.view(B, feature_dim, n).transpose(-1, -2)  # [B, n, feature_dim]
+    interp = interp.squeeze()  # [B?, n, feature_dim?]
+
+    return interp
+
+
+def interpolate_ms_features(
+    pts: torch.Tensor,
+    ms_grids: Union[nn.ModuleList, List[nn.ParameterList]],
+    grid_dimensions: int,
```
Comment (Giodiro) [inline nerfstudio/fields/kplanes_field.py:103]: Similarly `grid_dimensions` is always 2

File: nerfstudio/fields/kplanes_field.py [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+        grid_input = ray_samples.frustums.get_positions().view(-1, 3)
+
+        if self.spatial_distortion is not None:
+            grid_input = self.spatial_distortion(grid_input)
+            grid_input = grid_input / 2  # from [-2, 2] to [-1, 1]
+        else:
+            # Input should be in [-1, 1]
+            grid_input = (grid_input - self.aabb[0]) * (2.0 / (self.aabb[1] - self.aabb[0])) - 1.0
+
+        if self.grid_config[0]["input_coordinate_dim"] == 4:
```
Comment (Giodiro) [inline nerfstudio/fields/kplanes_field.py:271]: Time needs to be normalized: in nerfstudio datasets it's in [0, 1] but for interpolation it must be in [-1, 1] (do `timestamps = (timestamps * 2) - 1`)

File: nerfstudio/fields/kplanes_field.py [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+        grid_input = ray_samples.frustums.get_positions().view(-1, 3)
+
+        if self.spatial_distortion is not None:
+            grid_input = self.spatial_distortion(grid_input)
+            grid_input = grid_input / 2
+        else:
+            grid_input = (grid_input - self.aabb[0]) * (2.0 / (self.aabb[1] - self.aabb[0])) - 1.0
+
+        if ray_samples.times is not None and self.hexplane:
+            grid_input = torch.cat([grid_input, ray_samples.times.view(-1, 1)], -1)
```
Comment (Giodiro) [inline nerfstudio/fields/kplanes_field.py:398]: As above, time needs to be normalized

File: nerfstudio/models/kplanes.py [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+    _target: Type = field(default_factory=lambda: KPlanesModel)
+
+    near_plane: float = 2.0
+    """How far along the ray to start sampling."""
+    far_plane: float = 6.0
+    """How far along the ray to stop sampling."""
+
+    grid_config: List[Dict] = field(
+        default_factory=lambda: [
+            {"grid_dimensions": 2, "input_coordinate_dim": 3, "output_coordinate_dim": 32, "resolution": [64, 64, 64]}
```
Comment (Giodiro) [inline nerfstudio/models/kplanes.py:73]: here grid_dimensions is always 2, I would delete it.

File: nerfstudio/models/kplanes.py [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+        ]
+    )
+
+    is_ndc: bool = False
```
Comment (Giodiro) [inline nerfstudio/models/kplanes.py:77]: Never used

File: nerfstudio/models/kplanes.py [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+    is_contracted: bool = False
+    multiscale_res: List[int] = field(default_factory=lambda: [1, 2, 4, 8])
+    concat_features_across_scales: bool = True
+    linear_decoder: bool = False
+    linear_decoder_layers: Optional[int] = 4
```
Comment (Giodiro) [inline nerfstudio/models/kplanes.py:82]: default to 1 is better

File: nerfstudio/models/kplanes.py [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+            self.density_fns.extend([network.density_fn for network in self.proposal_networks])
+
+        # Samplers
+        update_schedule = lambda step: np.clip(
+            np.interp(step, [0, self.config.proposal_warmup], [0, self.config.proposal_update_every]),
+            1,
+            self.config.proposal_update_every,
+        )
+
+        if self.config.is_contracted or self.config.is_ndc:
```
Comment (Giodiro) [inline nerfstudio/models/kplanes.py:180]: Here there's no point in using the `is_ndc` config as nerfstudio doesn't have ndc datasets

File: nerfstudio/models/kplanes.py [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+
+        return outputs
+
+    def get_metrics_dict(self, outputs, batch):
+        image = self._get_image(batch, outputs)
+
+        metrics_dict = {}
+        metrics_dict["psnr"] = self.psnr(outputs["rgb"], image)
+        if self.training:
+            metrics_dict["plane_tv"] = compute_grid_tv(self.field.grids)
```
Comment (Giodiro) [inline nerfstudio/models/kplanes.py:296]: Maybe these should be computed in the `get_loss_dict` function directly (they're not really metrics like PSNR is).

**Review by Giodiro — COMMENTED** [review]
Hi @hturki , Thanks for the PR! I left a few minor comments on places which can be cleaned up; as well as the normalization of time which is important.
I'll try to add a commit on top for the missing regularizers!

**Comment by Giodiro** [comment]
To add to @iSach 's comment, you can have a look [here](https://github.com/akristoffersen/nerfstudio_kplanes/pull/5) for his implementation of the regularizers!

**Comment by iSach** [comment]
> Is there a reason you left behind the temporal losses? (sparse transients, temporal smoothness)
> 
> Also, I'm not sure I understand the code for the plane TV loss. In the paper, they apply it only on space-only planes and in 1D on space-time planes. However, the implementation here computes in 2D two times over space-only planes, and one time over space-time planes?
> 
> ```python
>     total = 0.0
>     for grids in multi_res_grids:
>         if len(grids) == 3:
>             spatial_grids = [0, 1, 2]
>         else:
>             spatial_grids = [0, 1, 3]  # These are the spatial grids; the others are spatiotemporal
>         for grid_id in spatial_grids:
>             total += compute_plane_tv(grids[grid_id])
>         for grid in grids:
>             total += compute_plane_tv(grid)
> ```

Actually, I don't even understand exactly what is going on in the "compute_plane_tv" function.

```
    batch_size, c, h, w = t.shape
    count_h = batch_size * c * (h - 1) * w
    count_w = batch_size * c * h * (w - 1)
    h_tv = torch.square(t[..., 1:, :] - t[..., : h - 1, :]).sum()
    w_tv = torch.square(t[..., :, 1:] - t[..., :, : w - 1]).sum()
    return 2 * (h_tv / count_h + w_tv / count_w)  # This is summing over batch and c instead of avg
```

* batch_size is always 1. Why is there this dimension? It's not the batch size, it's number of models.
* the sum() then dividing by count_h (or w) is exactly the same here as doing .mean() on h_tv (or w_tv). What is meant here by summing over batch and c instead of avg?
* Why is the result multiplied by 2?

If anybody understands this better, I'd like to know about it.

File: nerfstudio/configs/base_config.py [hunk 11]
```diff
@@ -28,6 +28,7 @@
 
 warnings.filterwarnings("ignore", module="torchvision")
 
+
```
[the remark this answers is not in the document]
Comment (hturki) [reply nerfstudio/configs/base_config.py:31]: Will do, but FYI I get these sort of extraneous diffs after running "black docs/ nerfstudio/ scripts/ tests/." I'll run it selectively only on the files that I touch but it would be nice if one could just run that one command and not have to check file-by-file in the future

File: nerfstudio/models/kplanes.py (hunk 10 above)
Comment (hturki) [reply nerfstudio/models/kplanes.py:296]: I personally like being able to look at these values before they get scaled with loss coefficients, but I don't have very strong opinions if others disagree

File: nerfstudio/configs/base_config.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (tancik) [reply nerfstudio/configs/base_config.py:31]: Running `ns-dev-test` should run black, isort, lint, tests

**Comment by hturki** [comment]
Higher-level comments:
- Will split out the alpha compositing into its own PR
- Re: some of the TV-related code - I mainly copied code from the official repo at https://github.com/sarafridov/K-Planes to make sure that the port was faithful / that the numbers line up with the paper. I'd refer any questions to the official implementation, but happy to simplify things / cut out parts that aren't relevant to nerfstudio
- Re: not adding time-related losses - I was hoping to start with that's needed for static scenes in this PR (which I should have mentioned in the PR description), as the paper mentions doing things like importance sampling for the DyNeRF scenes that would likely involve making more changes than what I wanted to do with this initial PR. I'm happy to add the time losses in this PR, but I (or someone else) would ideally do any other required things (like the importance sampling) in a follow-up to this

**Comment by Giodiro** [comment]
> Higher-level comments:
> 
>     * Will split out the alpha compositing into its own PR
> 
>     * Re: some of the TV-related code - I mainly copied code from the official repo at https://github.com/sarafridov/K-Planes to make sure that the port was faithful / that the numbers line up with the paper. I'd refer any questions to the official implementation, but happy to simplify things / cut out parts that aren't relevant to nerfstudio
> 
>     * Re: not adding time-related losses - I was hoping to start with that's needed for static scenes in this PR (which I should have mentioned in the PR description), as the paper mentions doing things like importance sampling for the DyNeRF scenes that would likely involve making more changes than what I wanted to do with this initial PR. I'm happy to add the time losses in this PR, but I (or someone else) would ideally do any other required things (like the importance sampling) in a follow-up to this

Regarding time: the PR already seems to support time-datasets, you can test that with dycheck or dnerf which are in nerfstudio. importance sampling is definitely out of scope though!

**Comment by iSach** [comment]
I indeed think importance sampling is out of scope, but it can be a very important convergence factor.

I have a large dataset (10500 images) with a large scale scene with small dynamic objects, and it made qualitative results incredibly better by implementing (some sort of) importance sampling.

However, my implementation changes more and bigger parts of Nerfstudio (and is probably not the cleanest), so I'll probably do it separately.

**Review by aaryapatel007 — APPROVED** [review]

File: nerfstudio/configs/base_config.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (hturki) [reply nerfstudio/configs/base_config.py:31]: my point is that running ns-dev-test causes changes in files that I'm not touching. For example, I see diffs such as this when running it on a fresh checkout of main:

image

Furthermore, running pylint (on a fresh checkout) seems to fail:

image

Is there possibly something I'm doing wrong here?

File: nerfstudio/engine/schedulers.py [hunk 12]
```diff
@@ -107,3 +108,37 @@ def func(step):
 
         scheduler = lr_scheduler.LambdaLR(optimizer, lr_lambda=func)
         return scheduler
+
+
+@dataclass
+class CosineWithWarmupSchedulerConfig(SchedulerConfig):
```
[the remark this answers is not in the document]
Comment (tancik) [reply nerfstudio/engine/schedulers.py:114]: Have you run `pip install -e .[dev]`?

File: nerfstudio/configs/base_config.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (tancik) [reply nerfstudio/configs/base_config.py:31]: Have you run `pip install -e .[dev]`?

[the remark this answers is not in the document]
Comment (hturki) [reply nerfstudio/configs/base_config.py:31]: hmm, I must have had the wrong versions of the libraries installed or something. things work now after reinstalling - sorry for the noise

**Review by aaryapatel007 — APPROVED** [review]

**Comment by hturki** [comment]
@tancik I think I've addressed most of the feedback in this PR, but I'm now getting a test failure (https://github.com/nerfstudio-project/nerfstudio/actions/runs/4423210082/jobs/7755716203) - do you know what might be going on?

**Review by aaryapatel007 — APPROVED** [review]

File: nerfstudio/fields/kplanes_field.py [hunk 13]
```diff
@@ -0,0 +1,388 @@
...
+    import tinycudann as tcnn
+except ImportError:
+    # tinycudann module doesn't exist
+    pass
+
+CONSOLE = Console(width=120)
+
+
+def init_grid_param(out_dim: int, reso: List[int], a: float = 0.1, b: float = 0.5):
+    """Initializes the grid parameters."""
```
Comment (tancik) [inline nerfstudio/fields/kplanes_field.py:46]: For all docstrings can you follow the form,

```
""" Desc

Args:
      a: stuff
      
Returns:
      desc
"""
```

File: nerfstudio/models/kplanes.py [hunk 14]
```diff
@@ -0,0 +1,420 @@
...
+    """K-Planes Model Config"""
+
+    _target: Type = field(default_factory=lambda: KPlanesModel)
+
+    near_plane: float = 2.0
+    """How far along the ray to start sampling."""
+    far_plane: float = 6.0
+    """How far along the ray to stop sampling."""
+
+    grid_config: List[Dict] = field(
```
Comment (tancik) [inline nerfstudio/models/kplanes.py:70]: Add argstrings to all arguments

**Review by tancik — COMMENTED** [review]
Thoughts on moving the k-planes encoding out of the field and into `encodings.py` similar to what we did with tensorf. This will make it easier for others to use these encodings in their projects.

**Comment by hturki** [comment]
@tancik I think we could move the k-planes encoding to ```encodings.py```, but there seems to be a ```TriplaneEncoding``` class there already for TensorRF that behaves differently from the K-Planes one and I'm not sure it would be straightforward to cleanly merge in the functionality needed to support both into that single encoding. Should I create a separate ```KPlanesEncoding``` or do you have a better suggestion?

**Comment by tancik** [comment]
> @tancik I think we could move the k-planes encoding to `encodings.py`, but there seems to be a `TriplaneEncoding` class there already for TensorRF that behaves differently from the K-Planes one and I'm not sure it would be straightforward to cleanly merge in the functionality needed to support both methods into that single encoding. Should I create a separate `KPlanesEncoding` or do you have a better suggestion?

Creating a `KPlanesEncoding` sounds good to me.

**Review by aaryapatel007 — APPROVED** [review]

**Comment by akristoffersen** [comment]
Is there something that's blocking this PR? Happy to get the branch updated / get through the final cleanup if no one else is working on it.

**Comment by Giodiro** [comment]
If we can merge the encoding first (https://github.com/nerfstudio-project/nerfstudio/pull/1658) and use it from this PR it should be good to go?

**Comment by hturki** [comment]
hey all, I've been dealing with a paper-related deadline but getting this over the line should be straightforward and I'll take care of it today

**Comment by hturki** [comment]
hey @tancik - are there any todos that you'd still like for me to address in this PR?

**Comment by tancik** [comment]
Sorry for the delay, Im traveling at the moment, won't have a chance to do a thorough look for a few days.

**Comment by hturki** [comment]
no rush on my side!

**Comment by tancik** [comment]
Sorry for taking so long to get back to this. Since my last comment we have started rethinking how we organize new models. Would you be willing to host the k-planes repo on your own github? We have a guide for setting it up [here](https://docs.nerf.studio/en/latest/developer_guides/new_methods.html), given this PR, it should slot in nicely. We can then add a page to methods so that others can easily find it. For example, we just did this for [instruct-nerf2nerf](https://docs.nerf.studio/en/latest/nerfology/methods/in2n.html)

**Comment by hturki** [comment]
sgtm, this will realistically happen near the end of may at the earliest but will let you know once I give it a go

**Comment by Giodiro** [comment]
Hi! K-planes integration is live at https://github.com/Giodiro/kplanes_nerfstudio
I assumed the k-planes embedding will still live in nerfstudio, but if you want to change that let me know!

I'll write a PR for the docs soon, and I think we can close this one!

**Comment by tancik** [comment]
> Hi! K-planes integration is live at https://github.com/Giodiro/kplanes_nerfstudio I assumed the k-planes embedding will still live in nerfstudio, but if you want to change that let me know!
> 
> I'll write a PR for the docs soon, and I think we can close this one!

This is great, thanks for your effort! We should keep the embedding in the main repo. Looking forward to the docs PR. Will close this PR.
